src/Eshopworld.WorkerProcess/WorkerLease.cs

```diff
@@ -80,45 +73,28 @@ public void StopLeasing()
 
         internal async Task LeaseAsync()
         {
-            _timer.Stop();
-
-            try
+            if (CheckLeaseExpired(CurrentLease))
             {
-                if (CheckLeaseExpired(CurrentLease))
-                {
-                    await _leaseAllocator.ReleaseLeaseAsync(CurrentLease).ConfigureAwait(false);
-
-                    OnLeaseExpired();
-                }
-
-                CurrentLease = await _leaseAllocator.AllocateLeaseAsync(InstanceId).ConfigureAwait(false);
-
-                if (CurrentLease != null)
-                    OnLeaseAllocated(CurrentLease.LeasedUntil.GetValueOrDefault());
-
-                if(CurrentLease?.LeasedUntil.HasValue ?? false)
-                {
-                    _timer.Interval = CurrentLease.Interval.Value.TotalMilliseconds;
-                }
-                else
-                {
-                    _timer.Interval = _slottedInterval.Calculate(
-                        CurrentLease?.LeasedUntil ?? ServerDateTime.UtcNow,
-                        _options.Value.LeaseInterval).TotalMilliseconds;
-                }
+                await _leaseAllocator.ReleaseLeaseAsync(CurrentLease).ConfigureAwait(false);
+
+                OnLeaseExpired();
             }
-            finally
+
+            CurrentLease = await _leaseAllocator.AllocateLeaseAsync(InstanceId).ConfigureAwait(false);
+
+            if (CurrentLease != null)
+                OnLeaseAllocated(CurrentLease.LeasedUntil.GetValueOrDefault());
+
+            if (CurrentLease?.LeasedUntil.HasValue ?? false)
             {
-                _timer.Start();
+                await _timer.ExecuteIn(CurrentLease.Interval.Value, LeaseAsync);
```

Comment:
Even when this implementation fixes the skew, the first parameter to ExecuteIn should still be the time left on the lease instead of the interval value - `(CurrentLease.LeaseUntil - ServerDateTime).Milliseconds`

Reply:
Can you cause the stackoverflow eventually here? - you are keep calling the same method every time and the stack seems to be staying unwound?

Reply:
good point thanks! moved the heavy lifting of the loop to the timer itself, while Lease async periodically updates the delay interval

Reply:
@nikhilgirraj addressed your comment as well